### PR TITLE
feat: introduce toolbar component

### LINKED ISSUE #2451: New Component - Toolbar
#### Is this a bug, enhancement, or feature request?
feature

#### Briefly describe your proposal.
introduce new component - Toolbar delivered in styles with [844](https://github.com/SAP/fundamental-styles/issues/840)

### PR DESCRIPTION
**Pull request by rengare**
#### Please provide a link to the associated issue.
Closes #2451 

#### Please provide a brief summary of this pull request.
The PR brings fiori3 toolbar component to ngx

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/master/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist

Documentation checklist:
- [x] update `README.md`
- [x] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [x] Documentation Examples
- [] Stackblitz works for all examples

### CODE REVIEW

**Comment by mikerodonnell89** [comment]
There is a difference between "info" state and "active" state in styles and they seem to be merged into the same thing here

File: apps/docs/src/app/core/component-docs/toolbar/examples/toolbar-overflow-example.component.html [hunk 1]
```diff
@@ -0,0 +1,37 @@
+<fd-toolbar [shouldOverflow]="true">
+    <label fd-toolbar-item></label>
```
Comment (mikerodonnell89) [inline apps/docs/src/app/core/component-docs/toolbar/examples/toolbar-overflow-example.component.html:2]: i think this can be removed

**Comment by mikerodonnell89** [comment]
I think we need directives to apply the classes `fd-toolbar__overflow__label` and `fd-toolbar__overflow__form-label` to stuff that goes in the overflow menu, see styles example

**Comment by mikerodonnell89** [comment]
Is there a way to ensure there is space between these two buttons with the overflow menu?  Tokenizer could be a good resource here

Screen Shot 2020-06-12 at 3 39 20 PM

**Comment by mikerodonnell89** [comment]
Screen Shot 2020-06-12 at 3 52 59 PM

fullscreen spacer isn't working

File: libs/core/src/lib/toolbar/toolbar.component.ts [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    QueryList,
+    HostListener,
+    ChangeDetectorRef,
+    Renderer2,
+    Input
+} from '@angular/core';
+import { ToolbarItemDirective } from './public_api';
+import { applyCssClass, CssClassBuilder } from '../utils/public_api';
+
+const ELEMENT_MARGIN = 8;
```
Comment (mikerodonnell89) [inline libs/core/src/lib/toolbar/toolbar.component.ts:21]: We should use getComputedStyle() to get these so we don't have to rely on constants that could potentially change in the future

https://stackoverflow.com/a/6338234

**Comment by mikerodonnell89** [comment]
Probably should add some margin to the overflow button to prevent this from happening.  Or take a look at how tokenizer handles collapsing stuff

Screen Shot 2020-06-12 at 3 39 20 PM

File: libs/core/src/lib/toolbar/toolbar.component.ts [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    @ViewChild('overflowBody')
+    overflowBody: ElementRef;
+
+    /** @hidden */
+    @ViewChild('overflowSpacer')
+    overflowSpacer: ElementRef;
+
+    /** @hidden */
+    @ContentChildren(forwardRef(() => ToolbarItemDirective))
+    toolbarItems!: QueryList<ToolbarItemDirective>;
```
Comment (mikerodonnell89) [inline libs/core/src/lib/toolbar/toolbar.component.ts:80]: Re: exclamation mark - I think there are certain cases when there might not be toolbar item children, for example if the toolbar just has a title

File: libs/core/src/lib/toolbar/toolbar.component.ts (hunk 2 above)
Comment (rengare) [reply libs/core/src/lib/toolbar/toolbar.component.ts:21]: The problem with getComputedStyle() is the performance, I tried that method and it's not fast

File: libs/core/src/lib/toolbar/toolbar.component.ts (hunk 3 above)
Comment (rengare) [reply libs/core/src/lib/toolbar/toolbar.component.ts:80]: leftover, nice catch :)

File: apps/docs/src/app/core/component-docs/toolbar/examples/toolbar-overflow-example.component.html (hunk 1 above)
Comment (rengare) [reply apps/docs/src/app/core/component-docs/toolbar/examples/toolbar-overflow-example.component.html:2]: fd-toolbar-item is needed because in overflow mode I query for all fd-toolbar-items

**Comment by rengare** [comment]
> Probably should add some margin to the overflow button to prevent this from happening. The "___ more" tokenizer might address this problem somehow, i can't remember
> 
> Screen Shot 2020-06-12 at 3 39 20 PM

I added extra space here

**Comment by rengare** [comment]
> Transparent toolbars need the 'fd-toolbar--clear' class to remove the bottom border, and it should be optional for auto toolbars
> 
> Screen Shot 2020-06-12 at 3 29 38 PM

I added property for that

File: apps/docs/src/app/core/component-docs/toolbar/examples/toolbar-overflow-example.component.html (hunk 1 above)
Comment (mikerodonnell89) [reply apps/docs/src/app/core/component-docs/toolbar/examples/toolbar-overflow-example.component.html]: There are other toolbar items, I'm confused, why do you need an empty one?

File: libs/core/src/lib/toolbar/toolbar.component.ts (hunk 3 above)
Comment (mikerodonnell89) [reply libs/core/src/lib/toolbar/toolbar.component.ts:88]: Still there, maybe forgot to stage for commit?

**Comment by mikerodonnell89** [comment]
Still missing the distinction between "info" and "active" state

**Comment by mikerodonnell89** [comment]
Resizing overflow menu should be able to move separators into the menu as well, like on styles:

Screen Shot 2020-06-15 at 10 13 57 AM

vs. here:
Screen Shot 2020-06-15 at 10 20 29 AM

**Comment by stefanoScalzo** [comment]
Need to fix the separator with the overflow. I shrank the page and it is not there

File: apps/docs/src/app/core/component-docs/toolbar/examples/toolbar-overflow-example.component.html (hunk 1 above)
Comment (rengare) [reply apps/docs/src/app/core/component-docs/toolbar/examples/toolbar-overflow-example.component.html]: ah right this one can be removed, I'll do it tomorrow

File: libs/core/src/lib/toolbar/toolbar.component.ts (hunk 3 above)
Comment (rengare) [reply libs/core/src/lib/toolbar/toolbar.component.ts]: I forgot to push changes, now its removed

**Comment by rengare** [comment]
> Still missing the distinction between "info" and "active" state

fixed

**Comment by rengare** [comment]
> Need to fix the separator with the overflow. I shrank the page and it is not there

fixed

**Comment by rengare** [comment]
> Resizing overflow menu should be able to move separators into the menu as well, like on styles:
> 
> Screen Shot 2020-06-15 at 10 13 57 AM
> 
> vs. here:
> Screen Shot 2020-06-15 at 10 20 29 AM

fixed

**Comment by mikerodonnell89** [comment]
This may be up for debate, in which case, I am happy moving forward with this PR and addressing my comments in another issue/PR.  But I wanted to point this out.

There is some unpredictable behavior with how items are removed from the toolbar and placed into the overflow container, and vice versa.  Copy this html to the overflow example

```
<fd-toolbar [shouldOverflow]="true">
    <button fd-toolbar-item fd-button [compact]="true">Button 1</button>
    <label fd-toolbar-label fd-toolbar-item>text 1</label>
    <fd-toolbar-separator fd-toolbar-item></fd-toolbar-separator>
    <button fd-toolbar-item fd-button [compact]="true">Button 2</button>
    <label fd-toolbar-label fd-toolbar-item>text 2</label>
    <fd-toolbar-separator fd-toolbar-item></fd-toolbar-separator>
    <button fd-toolbar-item fd-button [compact]="true">Button 3</button>
    <label fd-toolbar-label fd-toolbar-item>text 3</label>
    <button fd-toolbar-item fd-button [compact]="true">Button 4</button>
    <label fd-toolbar-label fd-toolbar-item>text 4</label>
    <fd-toolbar-spacer></fd-toolbar-spacer>
    <button fd-toolbar-item fd-button [compact]="true">Button 5</button>
    <label fd-toolbar-label fd-toolbar-item>text 5</label>
    <fd-toolbar-spacer></fd-toolbar-spacer>
    <fd-toolbar-separator fd-toolbar-item></fd-toolbar-separator>
    <button fd-toolbar-item fd-button [compact]="true">Button 6</button>
    <label fd-toolbar-label fd-toolbar-item>text 6</label>
    <fd-toolbar-separator fd-toolbar-item></fd-toolbar-separator>
    <button fd-toolbar-item fd-button [compact]="true">Button 7</button>
    <label fd-toolbar-label fd-toolbar-item>text 7</label>
    <button fd-toolbar-item fd-button [compact]="true">Button 8</button>
    <label fd-toolbar-label fd-toolbar-item>text 8</label>
    <fd-toolbar-spacer></fd-toolbar-spacer>
    <button fd-toolbar-item fd-button [compact]="true">Button 9</button>
    <label fd-toolbar-label fd-toolbar-item>text 9</label>
    <fd-toolbar-spacer></fd-toolbar-spacer>
    <fd-toolbar-separator fd-toolbar-item></fd-toolbar-separator>
    <button fd-toolbar-item fd-button [compact]="true">Button 10</button>
    <label fd-toolbar-label fd-toolbar-item>text 10</label>
    <fd-toolbar-separator fd-toolbar-item></fd-toolbar-separator>
    <button fd-toolbar-item fd-button [compact]="true">Button 11</button>
    <label fd-toolbar-label fd-toolbar-item>text 11</label>
    <button fd-toolbar-item fd-button [compact]="true">Button 12</button>
    <label fd-toolbar-label fd-toolbar-item>text 12</label>
    <fd-toolbar-spacer></fd-toolbar-spacer>
</fd-toolbar>
```

As you can see it goes button 1, text 1, button 2, text 2, etc.

In this gif I start resizing the window and you can see text 6 is hidden, then when button 6 gets hidden, text 6 is visible back in the toolbar.  I think items should be added to the overflow list sequentially, and what is happening here is that when button 6 is collapsed, this has freed up enough space for text 6 to come back, so it does.  I think in this case text 6 should remain in the overflow menu.  Otherwise, the order of items in the overflow menu does not correspond to how the application developer arranged them in the toolbar.

toolbarresize

Another example of this logic resulting in some weird behavior is here, where you will see multiple spacers next to each other in the toolbar, even though that's not how they're arranged in the HTML.

Screen Shot 2020-06-18 at 4 08 57 PM

**Comment by mikerodonnell89** [comment]
This may be up for debate, in which case, I am happy moving forward with this PR and addressing my comments in another issue/PR.  But I wanted to point this out.

There is some unpredictable behavior with how items are removed from the toolbar and placed into the overflow container, and vice versa.  Copy this html to the overflow example

```
<fd-toolbar [shouldOverflow]="true">
    <button fd-toolbar-item fd-button [compact]="true">Button 1</button>
    <label fd-toolbar-label fd-toolbar-item>text 1</label>
    <fd-toolbar-separator fd-toolbar-item></fd-toolbar-separator>
    <button fd-toolbar-item fd-button [compact]="true">Button 2</button>
    <label fd-toolbar-label fd-toolbar-item>text 2</label>
    <fd-toolbar-separator fd-toolbar-item></fd-toolbar-separator>
    <button fd-toolbar-item fd-button [compact]="true">Button 3</button>
    <label fd-toolbar-label fd-toolbar-item>text 3</label>
    <button fd-toolbar-item fd-button [compact]="true">Button 4</button>
    <label fd-toolbar-label fd-toolbar-item>text 4</label>
    <fd-toolbar-spacer></fd-toolbar-spacer>
    <button fd-toolbar-item fd-button [compact]="true">Button 5</button>
    <label fd-toolbar-label fd-toolbar-item>text 5</label>
    <fd-toolbar-spacer></fd-toolbar-spacer>
    <fd-toolbar-separator fd-toolbar-item></fd-toolbar-separator>
    <button fd-toolbar-item fd-button [compact]="true">Button 6</button>
    <label fd-toolbar-label fd-toolbar-item>text 6</label>
    <fd-toolbar-separator fd-toolbar-item></fd-toolbar-separator>
    <button fd-toolbar-item fd-button [compact]="true">Button 7</button>
    <label fd-toolbar-label fd-toolbar-item>text 7</label>
    <button fd-toolbar-item fd-button [compact]="true">Button 8</button>
    <label fd-toolbar-label fd-toolbar-item>text 8</label>
    <fd-toolbar-spacer></fd-toolbar-spacer>
    <button fd-toolbar-item fd-button [compact]="true">Button 9</button>
    <label fd-toolbar-label fd-toolbar-item>text 9</label>
    <fd-toolbar-spacer></fd-toolbar-spacer>
    <fd-toolbar-separator fd-toolbar-item></fd-toolbar-separator>
    <button fd-toolbar-item fd-button [compact]="true">Button 10</button>
    <label fd-toolbar-label fd-toolbar-item>text 10</label>
    <fd-toolbar-separator fd-toolbar-item></fd-toolbar-separator>
    <button fd-toolbar-item fd-button [compact]="true">Button 11</button>
    <label fd-toolbar-label fd-toolbar-item>text 11</label>
    <button fd-toolbar-item fd-button [compact]="true">Button 12</button>
    <label fd-toolbar-label fd-toolbar-item>text 12</label>
    <fd-toolbar-spacer></fd-toolbar-spacer>
</fd-toolbar>
```

As you can see it goes button 1, text 1, button 2, text 2, etc.

In this gif I start resizing the window and you can see text 6 is hidden, then when button 6 gets hidden, text 6 is visible back in the toolbar.  I think items should be added to the overflow list sequentially, and what is happening here is that when button 6 is collapsed, this has freed up enough space for text 6 to come back, so it does.  I think in this case text 6 should remain in the overflow menu.  Otherwise, the order of items in the overflow menu does not correspond to how the application developer arranged them in the toolbar.

toolbarresize

Another example of this logic resulting in some weird behavior is here, where you will see multiple spacers next to each other in the toolbar, even though that's not how they're arranged in the HTML.

Screen Shot 2020-06-18 at 4 08 57 PM

**Comment by mikerodonnell89** [comment]
Also that gif demonstrates an issue where the more button is removed from view as the resize is happening

**Comment by droshev** [comment]
> Also that gif demonstrates an issue where the more button is removed from view as the resize is happening

I would say let's try to fix that issue in this PR

**Comment by rengare** [comment]
> Also that gif demonstrates an issue where the more button is removed from view as the resize is happening

This problem shows because of spacer. In this case spacer takes remaining space which means when button is removed from content and moved to overflow, spacer takes button's space.
To be honest I don't know how to fix it

**Comment by mikerodonnell89** [comment]
I vote for merging this so we can use it in table and opening separate issues for the resizing strangeness

File: apps/docs/src/app/core/component-docs/toolbar/examples/toolbar-example.component.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+@Component({
+    selector: 'fd-toolbar-size-example',
+    templateUrl: './toolbar-size-example.component.html'
+})
+export class ToolbarSizeExampleComponent implements OnInit {
+    constructor() {}
+
+    ngOnInit() {}
```
Comment (JKMarkowski) [inline apps/docs/src/app/core/component-docs/toolbar/examples/toolbar-example.component.ts:60]: Let's remove empty constructor and ngOnInit

File: libs/core/src/lib/panel/panel.component.scss [hunk 5]
```diff
@@ -1,4 +1,4 @@
-@import '~fundamental-styles/dist/panel';
+// @import '~fundamental-styles/dist/panel';
 
 .fd-panel__content:focus {
     outline: none;
```
Comment (JKMarkowski) [inline libs/core/src/lib/panel/panel.component.scss:4]: We need this scss file.

File: libs/core/src/lib/toolbar/toolbar-spacer.component.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    @HostBinding('attr.class')
+    get css(): string {
+        return `fd-toolbar__spacer ${this.fixed ? 'fd-toolbar__spacer--fixed' : ''} ${this.class}`;
+    }
+
+    constructor() {}
+
+    ngAfterViewInit() {}
+
+    ngAfterViewChecked() {}
```
Comment (JKMarkowski) [inline libs/core/src/lib/toolbar/toolbar-spacer.component.ts:50]: Let's remove unused implementations

File: libs/core/src/lib/toolbar/toolbar.component.scss [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+}
+
+.fd-toolbar > * {
+    flex-shrink: 0;
+}
+
+// .fd-toolbar > .fd-popover {
+//     position: absolute;
+//     right: 0;
+// }
```
Comment (JKMarkowski) [inline libs/core/src/lib/toolbar/toolbar.component.scss:14]: Commented code

File: libs/core/src/lib/toolbar/toolbar.component.spec.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+    beforeEach(() => {
+        fixture = TestBed.createComponent(ToolbarComponent);
+        component = fixture.componentInstance;
+        fixture.detectChanges();
+    });
+
+    it('should create', () => {
+        expect(component).toBeTruthy();
+    });
```
Comment (JKMarkowski) [inline libs/core/src/lib/toolbar/toolbar.component.spec.ts:26]: Let's add tests for overflow behaviour and methods.

File: apps/docs/src/app/core/component-docs/toolbar/examples/toolbar-example.component.ts (hunk 4 above)
Comment (rengare) [reply apps/docs/src/app/core/component-docs/toolbar/examples/toolbar-example.component.ts]: done

File: libs/core/src/lib/panel/panel.component.scss (hunk 5 above)
Comment (rengare) [reply libs/core/src/lib/panel/panel.component.scss]: done

File: libs/core/src/lib/toolbar/toolbar-spacer.component.ts (hunk 6 above)
Comment (rengare) [reply libs/core/src/lib/toolbar/toolbar-spacer.component.ts]: done

File: libs/core/src/lib/toolbar/toolbar.component.scss (hunk 7 above)
Comment (rengare) [reply libs/core/src/lib/toolbar/toolbar.component.scss]: done

File: libs/core/src/lib/toolbar/toolbar.component.spec.ts (hunk 8 above)
Comment (rengare) [reply libs/core/src/lib/toolbar/toolbar.component.spec.ts:105]: done

**Comment by stefanoScalzo** [comment]
Screen Shot 2020-06-23 at 9 55 49 AM
There should be spacing here

**Comment by stefanoScalzo** [comment]
Add knobs for clear border, active and the text of the buttons in storybook

File: libs/core/src/lib/toolbar/toolbar.component.ts [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+    private _overflowElements: ToolbarItemDirective[] = [];
+
+    /** @hidden */
+    private _normalElements: ToolbarItemDirective[] = [];
+
+    /** @hidden */
+    constructor(private cd: ChangeDetectorRef, private renderer: Renderer2) {}
+
+    /** @hidden */
+    ngOnInit() {}
```
Comment (JKMarkowski) [inline libs/core/src/lib/toolbar/toolbar.component.ts:126]: Let's remove it

File: libs/core/src/lib/toolbar/toolbar.component.ts [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+    }
+
+    /** @hidden */
+    private _overflowElements: ToolbarItemDirective[] = [];
+
+    /** @hidden */
+    private _normalElements: ToolbarItemDirective[] = [];
+
+    /** @hidden */
+    constructor(private cd: ChangeDetectorRef, private renderer: Renderer2) {}
```
Comment (JKMarkowski) [inline libs/core/src/lib/toolbar/toolbar.component.ts:123]: Let's add `_` prefix

File: libs/core/src/lib/toolbar/toolbar.component.ts [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+    onResize() {
+        if (this.shouldOverflow) {
+            of(true)
+                .pipe(
+                    tap(() => this._reset()),
+                    delay(5),
+                    tap(() => this._collapseItems())
+                )
+                .subscribe(() => {});
+        }
```
Comment (JKMarkowski) [inline libs/core/src/lib/toolbar/toolbar.component.ts:158]: Great idea to put some delay there.

File: libs/core/src/lib/toolbar/toolbar.component.ts [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+        return toolbarItem.elementRef.nativeElement.offsetWidth + ELEMENT_MARGIN;
+    }
+
+    /** @hidden */
+    private _removeToolbarItemFromDOM(toolbarItem: ToolbarItemDirective) {
+        toolbarItem.elementRef.nativeElement.remove();
+    }
+
+    /** @hidden */
+    private _addToolbarItemToOverflow(toolbarItems: ToolbarItemDirective[]) {
```
Comment (JKMarkowski) [inline libs/core/src/lib/toolbar/toolbar.component.ts:220]: Can you add methods type return?

File: libs/core/src/lib/toolbar/toolbar.component.ts [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+        toolbarItems.forEach((x) => {
+            this._overflowBody.appendChild(x.elementRef.nativeElement);
+        });
+    }
+
+    /** @hidden */
+    private _reset() {
+        this._normalElements.map(this._removeToolbarItemFromDOM);
+        this._overflowElements.map(this._removeToolbarItemFromDOM);
```
Comment (JKMarkowski) [inline libs/core/src/lib/toolbar/toolbar.component.ts:229]: Any reasons why there is not forEach?

**Comment by mikerodonnell89** [comment]
@rengare if it is OK with you, I'd like to pick up this branch where you have currently left off, I am working on a solution to the flickering on resize

File: libs/core/src/lib/toolbar/toolbar.component.ts (hunk 12 above)
Comment (rengare) [reply libs/core/src/lib/toolbar/toolbar.component.ts:220]: done

File: libs/core/src/lib/toolbar/toolbar.component.ts (hunk 13 above)
Comment (rengare) [reply libs/core/src/lib/toolbar/toolbar.component.ts]: changed
